doc/man/8/rbd.rst

```diff
@@ -287,7 +287,15 @@ Commands
 :command:`encryption format` *image-spec* *format* *passphrase-file* [--cipher-alg *alg*]
   Formats image to an encrypted format.
   All data previously written to the image will become unreadable.
-  A cloned image cannot be formatted, although encrypted images can be cloned.
+  Image clones are inherently encrypted the same as their parent.
+  To format a cloned image, use `encryption format-thin`.
+  Supported formats: *luks1*, *luks2*.
+  Supported cipher algorithms: *aes-128*, *aes-256* (default).
+
+:command:`encryption format-thin` *image-spec* *format* *passphrase-file* *parent-format* *parent-passphrase-file* [--cipher-alg *alg*]
```

Comment:
As I understand you can't use `encryption format` for a clone and you can't use `encryption format-thin` for a non-clone. Then do we really need additional `format-thin` command (and additional API call)? Couldn't `encryption format` be extended to support additional options and be used for both cases?

Reply:
Correct.
Indeed my thought was to re-use encryption format for both cases.
However, it was Jason who insisted that we break it to separate API functions.
The thought was that we want the user to be aware that a thin-format does not erase up existing data (which belongs to the parent), as opposed to regular format which causes all old data to become unreadable.

Reply:
Ack. I would still prefer one command, but ok, I am fine with it.

Reply:
Rethinking about this:
How about we eliminate `encryption format-thin` CLI, and instead add `--parent-encryption-format` etc. to `rbd clone` CLI?
This way we guarantee that a single format operation is applied per clone. (re-formating flat images is legit, but re-formatting clones is not), as well as we guarantee that the format is applied immediately after the clone is created, and not prior to any writes being made to the clone (which will also cause corruption).

Reply:
Right now I don't see problems with this approach and actually like it more.

But I might be missing some background here. For example, I was surprised to learn that one had explicitly call `rbd encryption format` command after image creation, instead of just providing encryption options to `rbd create` command. There were probably some reasons for this which I am not aware?

Reply:
Well, for the C/C++ API we have encryption_format and encryption_load functions.
I didn't touch rbd_create and rbd_open, since these should stay as simple as possible.
And anyhow, we want to be able to also format/load on an image context. independently of create/open.

Now for the CLI, I felt rushed for the pacific freeze deadline, so I just added the minimum work which is to wrap encryption format and load to their own CLI.
As I noted, in any case we want to support these separate APIs.

Now there is the question I raised now on whether we extend rbd create/clone CLI with encryption.
I have not discussed this with Jason at all.
This is something I thought about just recently.

Reply:
To make `rbd create` work we would not need even to extend the API, just call create and then format, right?

But probably extending `create` API just would mean adding new image options: `RBD_IMAGE_OPTION_ENCRYPTION_FORMAT` and `RBD_IMAGE_OPTION_ENCRYPTION_PASSPHRASE`?
Which would also be used for `clone`.

Reply:
> To make `rbd create` work we would not need even to extend the API, just call create and then format, right?
> 

Right, that was my thought.

> But probably extending `create` API just would mean adding new image options: `RBD_IMAGE_OPTION_ENCRYPTION_FORMAT` and `RBD_IMAGE_OPTION_ENCRYPTION_PASSPHRASE`?
> Which would also be used for `clone`.

Well, there's also CIPHER_ALG when formatting.
While passphrase is simply a string, cipher alg and format are enums.
I didn't see any such other image options.

Note that other than encryption format, all other encryption parameters are format-specific (currently LUKS1/2),
but we should think about other formats for future compatibility. 
This is why I created `encryption_luks1_format_options_t`, `encryption_luks2_format_options_t`.

Given all that, do you still think we should extend the C/C++ API by adding new image options?

Reply:
If you can do without extending API, then it is good.

I just suspected we would need to do this anyway earlier or later, and the for create/clone I suggested just add image options. Actually you may have just one option, `RBD_IMAGE_OPTION_ENCRYPTION`, and encode here all encryption options you would need, e.g. as a json string.